### PR TITLE
Add ERCOT  System Lambda

### PR DESCRIPTION
**Pull request by Michael-fore**
#269 

For some reason I decided to work on this one first. 

I added a function to pass through the self.read_doc to override the columns renaming of the parse_doc
```
        def rename_func(df):
            df[["DeliveryDate", "HourEnding"]] = df["SCEDTimeStamp"].str.split(
                " ",
                expand=True,
            )
            return df

        df = self.read_doc(doc=doc, rename_func=rename_func)
```
and
```
if rename_func:
            doc = rename_func(doc)
else:
            doc.rename( ...)
```

Let me know if you don't want to use this approach

Excited for my first contribution!

#

### CODE REVIEW

File: gridstatus/ercot.py
Comment (kmax12) [inline gridstatus/ercot.py:1780]: we can use `@support_date_range("DAY_START")` as is if we use the get document approach I mention below

File: gridstatus/ercot.py
Comment (kmax12) [inline gridstatus/ercot.py:1797]: let's make this function return a full day of data if `date != 'latest'`

in that case where `date != latest`, we can just pass the date along to use self._get_documents instead and it will return a list of documents for that day

you can then parse and concatenate them

File: gridstatus/ercot.py
Comment (kmax12) [inline gridstatus/ercot.py:1803]: does it make sense for this dataset to get parsed to an interval start and interval end? 

it seems to me like it would be fine to just leave the columns as`SCED Time Stamp` and `System Lambda` for now. if we do want to do parse to intervals, we'll need some logic like you see [here](https://github.com/kmax12/gridstatus/blob/main/gridstatus/ercot.py#L1143). Basically, this converts the sced time stamp to the 15 minute spp interval. 

in either case,  you'll have to update `read_doc` function or just create a new one, since it doesn't really handle this well. right now. 

We will face a similar question when we integrate the real time 5 minute LMPs (we currently only do SPP).

**Review by kmax12 — CHANGES_REQUESTED** [review]
thanks for the contribution. here are some initial comments

File: gridstatus/ercot.py
Comment (Michael-fore) [reply gridstatus/ercot.py:1803]: I am so down to not  deal with interval ending if we don't need to 

IIRC this is an interval ending data type, but we can reach out to ERCOT to ask if need be

File: gridstatus/ercot.py
Comment (Michael-fore) [reply gridstatus/ercot.py:1797]: Is speed an issue you solve for?

As of now, fetching a whole day takes awhile.

I can speed it up if by batch fetching, but only want to do that if necessary

Comment (kmax12) [reply gridstatus/ercot.py:1797]: i think supporting a whole day and latest is enough for now. in theory it be nice to be able to specify a finer granularity, but that would require some reworking to get_documents. i would totally welcome it, but maybe out of scope

File: gridstatus/ercot.py
Comment (kmax12) [reply gridstatus/ercot.py:1803]: let's just keep it as non interval ending for now and we can also enhance later

File: gridstatus/ercot.py
Comment (kmax12) [inline gridstatus/ercot.py:1812]: small thing, it's more efficient to rewrite this do just one concat

```python
dfs = []
for doc in docs:
    log(f"Reading {doc.url}", verbose)
    new_df = pd.read_csv(doc.url, compression="zip")
    dfs.append(new_df)
df = pd.concat(dfs)
```

File: gridstatus/ercot.py
Comment (kmax12) [inline gridstatus/ercot.py:1802]: maybe just do `docs = [doc]` here and then have shared logic for reading the csv

File: gridstatus/ercot.py
Comment (Michael-fore) [reply gridstatus/ercot.py]: Just went the easy route on this one

File: gridstatus/ercot.py
Comment (kmax12) [inline gridstatus/ercot.py:1806]: i think making this plural variable called `docs` make more sense

File: gridstatus/ercot.py
Comment (kmax12) [inline gridstatus/ercot.py:1814]: this appears to be missing handle of the date time parsing, dst changes, and setting of the time zone

for example: https://github.com/kmax12/gridstatus/blob/main/gridstatus/ercot.py#L2028

if you can modify parse_doc function you maybe be able to inherit some of this logic for free

**Review by kmax12 — CHANGES_REQUESTED** [review]

File: gridstatus/ercot.py
Comment (Michael-fore) [reply gridstatus/ercot.py:1802]: I usually prefer more verbose logic, but changed as well

File: gridstatus/ercot.py
Comment (Michael-fore) [reply gridstatus/ercot.py:1812]: nice catch, fixed

File: gridstatus/ercot.py
Comment (Michael-fore) [reply gridstatus/ercot.py:1814]: Oh i thought we were good to ignore that, i may just redo the rename_func logic to inherit that logic

**Comment by Michael-fore** [comment]
Still fixing dst stuff

**Comment by Michael-fore** [comment]
@kmax12 Do you also shift the short day hour? So from 134 to 124?

File: gridstatus/ercot.py
Comment (Michael-fore) [reply gridstatus/ercot.py]: I uh, missed this one. lmk if you want it fixed

File: gridstatus/ercot.py
Comment (kmax12) [inline gridstatus/ercot.py:1825]: i think it's better to avoid numpy here and just use the pandas version, which i think is sometihng like this 

```
df["SCEDTimeStamp"] = df.apply(lambda row: handle_dst(row["SCEDTimeStamp"], row["RepeatedHourFlag"]), axis=1)
```

Comment (kmax12) [reply gridstatus/ercot.py:1825]: scratch that, i would just do this 

df["SCEDTimeStamp"] = pd.to_datetime(df["SCEDTimeStamp"]).dt.tz_localize(self.default_timezone, ambiguous=df["RepeatedHourFlag"] == "Y")

File: gridstatus/ercot.py
Comment (kmax12) [inline gridstatus/ercot.py:1830]: to be consistent with how we name variables, we want the final column names to be "SCED Time Stamp" and "System Lambda". Drop repeated hour since the time stamp column is properly typed and has timezone info

File: gridstatus/ercot.py
Comment (kmax12) [reply gridstatus/ercot.py:1825]: actually, can you check the logic of how the ambiguious param works and the repeated hour flag? 

unclear to me if RepeatedHourFlag is Y or N when it is DST. maybe need to be ambiguous=df["RepeatedHourFlag"] == "N"

here's the documentation: https://pandas.pydata.org/docs/reference/api/pandas.Series.tz_localize.html

Comment (Michael-fore) [reply gridstatus/ercot.py:1825]: from the docs

```bool-ndarray where True signifies a DST time, False designates a non-DST time (note that this flag is only applicable for ambiguous times)```

I believe those with `"Y"` are the ambiguous times, but we can also wait to hear back from ERCOT

Comment (kmax12) [reply gridstatus/ercot.py]: based on our slack discussion, i thought we settled that it is set to Y when the time repeats itself during the transition from DST to Standard Time, which would mean that the value of N represents DST. 

but yes, i have a question into ercot and let's see what they say

Comment (stephennimmo) [reply gridstatus/ercot.py]: Using an hourly example, when there is a DST event, you'll have 23 hourly events or 25 hourly events. In the xml, that Y flag denotes the event that is a duplicate based on the timestamps. Your assumption is correct. 

FWIW, I've typically seen this implemented by getting rid of "timezone on write", but rather storing the data as utc and then applying timezone on read. Then, you could just use Unix-epoch milliseconds as a format and store the appropriate timezone info in some sort of metadata. However, there are pros/cons to that as well like what happens with timezones change (they do) or having to do timezone conversions 1000s of times for every historical read. 

I've also seen the records flattened, where there is a single row for each day, with a column for each hourly ending value with the column name being the time sans punctuation. To support DST, in the hourly example, you would have a row with a date column followed by 25 potential hourly columns (0000, 0100, 0200, etc). This look great from a storage efficiency because now you are eliminating a bunch of repetitive data however if you want to read a specific column, you'll end up having to write some query magic and would at a minimum typically be reading a full daily row back.

File: gridstatus/ercot.py
Comment (kmax12) [inline gridstatus/ercot.py:2036]: i think you can back out these changes

File: gridstatus/ercot.py
Comment (kmax12) [inline gridstatus/ercot.py:1811]: to make this easier to use with backfilling data that is sent over directly by ERCOT, can you separate out the logic for handling documents into a _handle_sced_system_lambda function? you can see an example of this here: https://github.com/kmax12/gridstatus/blob/main/gridstatus/ercot.py#L1889

File: gridstatus/ercot.py
Comment (kmax12) [inline gridstatus/ercot.py:1820]: the final column names should be just: "SCED Time Stamp", "System Lambda". I dont think that is what this code does or what you are testing for

File: gridstatus/ercot.py
Comment (kmax12) [inline gridstatus/ercot.py:1815]: per our voluminous other discussion I think this should be 

`ambiguous=df["RepeatedHourFlag"] == "N"`

the repeated hour happens during standard time. we want ambiguous to be true during DST time.

File: gridstatus/tests/test_ercot.py
Comment (kmax12) [inline gridstatus/tests/test_ercot.py:13]: fix this test to make sure it passes. add test for  date="today" in addition to date='latest'

**Review by kmax12 — CHANGES_REQUESTED** [review]
no need to wait for response from ercot to merge this in. left some more comments and it should be ready to merge once they are addressed

**Comment by Michael-fore** [comment]
Changes implemented, if not this push then the next we should be good

File: gridstatus/ercot.py
Comment (kmax12) [inline gridstatus/ercot.py:1831]: sorry, just realized this in my final review. we normal return data in chronologically ascending order. 

can you add a `df.sort_values("SCED Time Stamp")`?

File: gridstatus/ercot.py
Comment (kmax12) [inline gridstatus/ercot.py:1814]: this should take a `verbose` parameter. 

before reading the URL, add logging statement.

also wrap the loop in tqdm for a progress bar. helping since a day downloads many files 

here is an example: https://github.com/kmax12/gridstatus/blob/main/gridstatus/ercot.py#L950

**Review by kmax12 — CHANGES_REQUESTED** [review]
just a couple more comments. the last thing is can you add a line item to `CHANGELOG.md`? 

before you do that, I recommend merging in `main` since I already created an entry for the next release

**Comment by Michael-fore** [comment]
re merged to main and implemented the  asks

File: gridstatus/ercot.py
Comment (kmax12) [inline gridstatus/ercot.py:1825]: tests arent passing because this is missing the required `verbose` parameter. let me know if you need help setting up tests to run. 

ill run it on github action, but unfortunately its current broken due to an unrelated issue of requests to SPP's server not working on github

Comment (Michael-fore) [reply gridstatus/ercot.py:1825]: Oh i have no idea how to set up those tests. Help would be nice

Comment (Michael-fore) [reply gridstatus/ercot.py]: does spp block the github server or something?

Comment (Michael-fore) [reply gridstatus/ercot.py]: verbose added tho

Comment (kmax12) [reply gridstatus/ercot.py]: start here for tests: https://github.com/kmax12/gridstatus/blob/main/CONTRIBUTING.md#running-tests-and-linting

Comment (kmax12) [reply gridstatus/ercot.py]: i think the problem with SPP has to do with SSL cert configuration on the docker image used for the github action. i need to investigate more

File: gridstatus/ercot.py
Comment (kmax12) [inline gridstatus/ercot.py:1831]: you dont need this, `if verbose`. log already handles it

**Comment by Michael-fore** [comment]
failing from gridstatus/tests/test_spp.py::TestSPP::test_get_lmp_historical[Markets.DAY_AHEAD_HOURLY]

**Review by kmax12 — APPROVED** [review]
Looks good and ready to merge. Thanks for contributions!

**Pull request merged**
